src/Clash/Cores/Ethernet/Mac/Preamble.hs

```diff
@@ -0,0 +1,60 @@
+{-# OPTIONS_HADDOCK hide #-}
+
+{-|
+Module      : Clash.Cores.Ethernet.Mac.Preamble
+Description : Provides components which insert and strip the Ethernet preamble.
+-}
+module Clash.Cores.Ethernet.Mac.Preamble (
+  preambleInserterC,
+  preambleStripperC,
+) where
+
+import Clash.Prelude
+
+import Protocols ((|>), Circuit)
+import Protocols.PacketStream
+
+-- | Ethernet start frame delimiter (SFD), least significant bit first.
+startFrameDelimiter :: BitVector 8
+startFrameDelimiter = 0xD5
+
+-- | The size of the Ethernet preamble + SFD.
+type Preamble = Vec 8 (BitVector 8)
+
+-- | The actual preamble, each byte ordered least significant bit first.
+preamble :: Preamble
+preamble = replicate d7 0x55 :< 0xD5
+
+{- |
+Prepends the Ethernet preamble and SFD to each packet in the packet stream.
+The bytes are ordered least significant bit first:
+>>> preamble = 0x55 :> 0x55 :> 0x55 :> 0x55 :> 0x55 :> 0x55 :> 0x55 :> 0xD5 :> Nil
+
+Inherits latency and throughput from `packetizerC`.
+-}
+preambleInserterC ::
+  forall dataWidth dom.
+  (HiddenClockResetEnable dom) =>
+  (KnownNat dataWidth) =>
+  (1 <= dataWidth) =>
+  Circuit (PacketStream dom dataWidth ()) (PacketStream dom dataWidth ())
+preambleInserterC = packetizerC id (const preamble)
+
+{- |
+Strips the incoming packet stream of the Ethernet preamble and SFD (Start Frame
+Delimiter), 8 bytes in total. Drops a packet if the SFD is not correct.
+Does not check if the preamble itself matches for efficiency reasons.
+
+Inherits latency and throughput from `depacketizerC`.
+-}
+preambleStripperC ::
+  forall dataWidth dom.
+  (HiddenClockResetEnable dom) =>
+  (KnownNat dataWidth) =>
+  (1 <= dataWidth) =>
+  Circuit (PacketStream dom dataWidth ()) (PacketStream dom dataWidth ())
+preambleStripperC =
```

Comment:
I'm fairly sure a receiver can discard the preamble; not sure whether in whole or if there is a minimum number of bit times it needs to pass on for valid reception. I've quickly browsed through 802.3 and found the following things:

802.3-2018 section 4.2.5 _Preamble generation_
> In a LAN implementation, most of the Physical Layer components are allowed to provide valid output some
number of bit times after being presented valid input signals. Thus it is necessary for a preamble to be sent
before the start of data, to allow the PLS circuitry to reach its steady state.

I think this says a receiver is allowed to only start outputting the preamble _some number of bit times_ after the preamble starts, and the preamble is there to absorb the difference.

4.1.2.1.2 _Reception without contention_
> The Physical Layer passes subsequent bits up to the MAC sublayer, where the leading bits are discarded, up to and including the end of the preamble and Start Frame Delimiter.

Not really conclusive, but I feel it leaves the door wide open to say the number of leading bits is inconsequential.

In section 4.2.9 _Frame reception_, IEEE 802.3 provides pseudo-Pascal programs to show what a working implementation of a CSMA/CD Media Access sublayer could look like. In it's `BitReceiver` process, which produces received frames, it strips off stuff by invoking another procedure; the call is documented:
```pascal
PhysicalSignalDecap; {Skip idle and extension, strip off preamble and sfd}
```
The procedure itself is only described:
```pascal
procedure PhysicalSignalDecap;
begin
{Receive one bit at a time from physical medium until a valid sfd is detected, discard bits and return}
end; {PhysicalSignalDecap}
```

Nowhere does it say anything about a check on the length of the preamble in the receiver.

Without referring to any standard, it also just makes sense. The preamble is, among other things, meant for synchronisation. In the first part of the preamble, you haven't synchronised yet, so you can't output it because it is still gibberish in some sense.

Reply:
I wrote two new versions of the preamble stripper that allow for a variable-length preamble in bytes. I believe it is reasonable to assume that the SFD is byte-aligned.

The first version just forwards fragments upon encountering the SFD:

```Haskell
data PreambleStripperState
  = ValidateSfd
  | Forward
  deriving (Generic, NFDataX, Show, ShowX)

preambleStripperC ::
  forall dom.
  HiddenClockResetEnable dom =>
  Circuit (PacketStream dom 1 ()) (PacketStream dom 1 ())
preambleStripperC = fromSignals (mealyB go ValidateSfd)
 where
  go ValidateSfd (Just PacketStreamM2S{..}, _) =
    (nextSt, (PacketStreamS2M True, Nothing))
   where
    nextSt
      | isNothing _last && head _data == 0xD5 = Forward
      | otherwise = ValidateSfd
  go Forward (Just transferIn, bwdIn) = (nextSt, (bwdIn, Just transferIn))
   where
    nextSt
      | isJust (_last transferIn) && _ready bwdIn = ValidateSfd
      | otherwise = Forward
  go st (Nothing, _) = (st, (PacketStreamS2M True, Nothing))
```
This version happily accepts packets with a longer preamble than normal just like the procedure described in the comment above. And prays that faulty packets are picked up by the CRC check.

Or we can go with a version that's a little stricter (but also slightly more expensive):

```Haskell
data PreambleStripperState
  = ValidateSfd {_counter :: Index 8}
  | Forward
  | Drop
  deriving (Generic, NFDataX, Show, ShowX)

preambleStripperC ::
  forall dom.
  HiddenClockResetEnable dom =>
  Circuit (PacketStream dom 1 ()) (PacketStream dom 1 ())
preambleStripperC = fromSignals (mealyB go (ValidateSfd 0))
 where
  go ValidateSfd{..} (Just PacketStreamM2S{..}, _) =
    (nextSt, (PacketStreamS2M True, Nothing))
   where
    nextSt
      | isNothing _last && head _data == 0xD5 = Forward
      | isNothing _last && _counter == maxBound = Drop
      | otherwise = ValidateSfd (satSucc SatWrap _counter)
  go Forward (Just transferIn, bwdIn) = (nextSt, (bwdIn, Just transferIn))
   where
    nextSt
      | isJust (_last transferIn) && _ready bwdIn = ValidateSfd 0
      | otherwise = Forward
  go Drop (Just transferIn, _) = (nextSt, (PacketStreamS2M True, Nothing))
   where
    nextSt
     | isJust (_last transferIn) = ValidateSfd 0
     | otherwise = Drop
  go st (Nothing, _) = (st, (PacketStreamS2M True, Nothing))
```

Which drops packets if the SFD was not detected in the first 8 bytes.

I personally prefer the latter, what do you guys think?

Reply:
I feel more for the former. I don't feel the extra circuitry adds a useful enough feature.

Also, in the latter, it seems to me you can replace `satSucc SatWrap _counter` just by `succ _counter`. It always starts at 0 and never progresses beyond `maxBound`.

Reply:
I don't really care. The hardware difference will probably be relatively tiny. But since this will run in the native ethernet domain which potentially very high speed it might be better to got for version 1.

Reply:
Then we will go with the former. My initial thoughts were that the former version could let packets that are too small through, which could break things like `dropTail` and others. But now I realise it doesn't really protect against that, and depacketizers will drop those faulty packets along the way anyway.

Reply:
> I believe it is reasonable to assume that the SFD is byte-aligned.

That sounds like a reasonable assumption for 1000BASE-X and 1000BASE-T. But I don't know if it's valid for 10BASE-T and 100BASE-T, which we also want to support. Heck, I'd like to support any not too obscure Ethernet. 10BASE-T uses Manchester encoding and 100BASE-T uses 4B5B.

Let's take a look. Thinking about Manchester encoding, I think if MII wants to nibble-align to the SFD and needs a number of preamble transitions to synchronise, I think it necessarily has to omit all of the preamble because there is no way to find nibble boundaries until you see the first nibble of the SFD. The preamble is just alternating 0's and 1's, you can't see nibble boundaries like you often (but not always) _can_ with 4B5B or 8B10B. So I expect 10BASE-T to by byte-aligned by virtue of not emitting preamble at all, but I don't know for sure.

Then 4B5B. 100BASE-T doesn't do preambles; it does IDLE code-groups between packets. These are all identical. I don't see a requirement that the number of IDLE nibbles between frames is even. The 100BASE-T standard is based on the FDDI standard and I don't have that one, so 802.3 is less conclusive. Still, it seems to me, if you're idling and you have a packet to transmit, you don't have to wait for an even number of IDLE nibbles to have passed. _However_! 802.3 says that the MII representation of the IDLE code-group is undefined! I think this necessarily means that RX_DV cannot be asserted during IDLE code-groups, and thus, that it is only asserted upon reception of the SFD, and it is byte-aligned.

Still, that's only 100BASE-T. The MII specification clearly defines preamble patterns even at the MII level. Section 22.2.3.2.2 of 802.3-2018 (_Receive case_) seems to be saying you either receive the whole preamble or no preamble. But I wouldn't bet on my interpretation being right.

A relevant portion of IEEE 802.3-2018 for MII seems section 22.2.2.7 _RX_DV (Receive Data Valid)_:

> RX_DV (Receive Data Valid) is driven by the PHY to indicate that the PHY is presenting recovered and
decoded nibbles on the RXD<3:0> bundle and that the data on RXD<3:0> is synchronous to RX_CLK.
RX_DV shall transition synchronously with respect to the RX_CLK. RX_DV shall remain asserted
continuously from the first recovered nibble of the frame through the final recovered nibble and shall be
negated prior to the first RX_CLK that follows the final nibble. In order for a received frame to be correctly
interpreted by the Reconciliation sublayer and the MAC sublayer, RX_DV must encompass the frame,
starting no later than the Start Frame Delimiter (SFD) and excluding any End-of-Frame delimiter.
Figure 22–7 shows the behavior of RX_DV during frame reception.

Nowhere does it say RX_DV can only transition to enabled state on a byte-aligned boundary.

I think it's all not very clear, and I only looked at twisted pair, but we want to support other physical layers as well. So I think you shouldn't rely on the SFD being byte-aligned. It might always be the case for several physical layers, but that still leaves open the possibility it is not the case for some other physical layer.

Reply:
How would you sensibly handle a non byte-aligned SFD with a streaming protocol that only allows you to enable full bytes? We cannot say that some bits are invalid. This is not possible with AXI Stream either.

Reply:
> How would you sensibly handle a non byte-aligned SFD with a streaming protocol that only allows you to enable full bytes?

I did not know about which place in the stack you were talking about. MII is a nibble-width protocol, and it's somewhat of a common unit in Ethernet. I thought you meant whether the nibbles of the SFD were byte-aligned, i.e., that the first nibble is always at an even number of cycles from the first where RX_DV is asserted (0 cycles, i.e., first nibble is the SFD, or 2, or ...).

It's a pity I misunderstood you. I spent a fair amount of time wading through 802.3 to check your assumption; turns out my assumption was the faulty one.

If I understand Rowan correctly, I agree that it should be byte-aligned before you turn it into a `PacketStream`. It doesn't seem to make sense to not do this.

Doesn't RGMII strictly define byte alignment? I.e., they either say the lower nibble is on the falling clock edge and the upper nibble is on the rising edge, or they say the exact reverse but still fully specify it? I'm not going to dive into another standard right now...